pkg/postage/service.go

```diff
@@ -165,24 +143,35 @@ func (ps *service) IssuerUsable(st *StampIssuer) bool {
 func (ps *service) GetStampIssuer(batchID []byte) (*StampIssuer, func() error, error) {
 	ps.lock.Lock()
 	defer ps.lock.Unlock()
-	for _, st := range ps.issuers {
-		if bytes.Equal(batchID, st.data.BatchID) {
-			if !ps.IssuerUsable(st) {
-				return nil, nil, ErrNotUsable
-			}
-			return st, func() error {
-				return ps.save(st)
-			}, nil
-		}
+
+	if _, ok := ps.issuerLocks[string(batchID)]; !ok {
+		ps.issuerLocks[string(batchID)] = &sync.Mutex{}
+	}
+	ps.issuerLocks[string(batchID)].Lock()
```

Comment:
So the problem with using sync.Mutex here is that it is a blocking call. This will block the uploader thread. If we want to make this a blocking call we need to use the context as an argument and then use some sort of context-based locking for eg x/sync.semaphore and pass the http.Request context so that on cancellation from user it exits cleanly. This will complicate things as you will be waiting while holding the service.Lock which can cause other problems.

I think its better to have an empty struct in the map just to signal that the batch ID is in use and if it exists we return error, something is ErrBatchInUse and we can say in docs that currently only 1 upload is allowed per batch.